tests/experimental/wgit/test_signal_sparsity.py

```diff
@@ -6,50 +6,63 @@
-    and compares it with the sst from dense_to_sst method."""
-
-    def assert_and_get_sps_sst(tensor, percent, element, dim):
-        dense_freq = torch.fft.fft(tensor)
-
-        real_dense_freq = dense_freq.real
-        sparser_2d = SignalSparsity(
-            sst_top_k_percent=percent, sst_top_k_element=element, sst_top_k_dim=dim, dst_top_k_percent=100
-        )
-        top_k_total_size = _top_k_total_size(tensor, dim)
-        k = _get_k_for_topk(percent, element, top_k_total_size)
-        default_sst = torch.zeros_like(dense_freq)
-
-        orig_shape = dense_freq.shape
-        if dim is None and len(orig_shape) > 1:
-            default_sst = default_sst.reshape(-1)
-            real_dense_freq = real_dense_freq.reshape(-1)
-            dense_freq = dense_freq.reshape(-1)
-            dim = -1
-
-        _, i = real_dense_freq.abs().topk(k, dim=dim)
-        default_sst = default_sst.scatter(dim, i, dense_freq.gather(dim, i)).reshape(orig_shape)
-
-        # verify if sst only has top-k values across the right dim
-        dim = None if (dim == -1) else dim
-        assert all((default_sst.abs().count_nonzero(dim) == k).flatten())
-
-        sst = sparser_2d.dense_to_sst(tensor)
-        assert all((default_sst == sst).flatten())
-        assert sst.shape == tensor.shape
-        return default_sst, sst
-
-    return assert_and_get_sps_sst
+from fairscale.experimental.wgit.signal_sparsity import SignalSparsity
+
+
+def expected_io_vals():
+    """Helper function to create and return a list of tuples of the form:
+    (in_tensor, expected_tensor, dim, percent) to be used as parameters for tests. The top-k
+    value for the expected SST was set at 2.
+    """
+    # Expected SST output for 4x3 tensor of ascending ints
+    expected_4x3_None = torch.tensor(
+        [
+            [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],  # with dim=None, top-2
+            [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
+            [21.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
+            [30.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
+        ],
+        dtype=torch.complex64,
+    )
+
+    expected_4x3_0 = torch.tensor(
+        [
+            [0.0000000 + 0.0000000j, 0.0000000 + 0.0000000j, 0.0000000 + 0.0000000j],  # with dim=0, top-2
+            [0.0000000 + 0.0000000j, 0.0000000 + 0.0000000j, 0.0000000 + 0.0000000j],
+            [21.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, -1.5000000 - 0.8660254j],
+            [30.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, -1.5000000 - 0.8660254j],
+        ],
+        dtype=torch.complex64,
+    )
+
+    expected_4x3_1 = torch.tensor(
+        [
+            [3.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, 0.0000000 + 0.0000000j],  # with dim=1, top-2
+            [12.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, 0.0000000 + 0.0000000j],
+            [21.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, 0.0000000 + 0.0000000j],
+            [30.0000000 + 0.0000000j, -1.5000000 + 0.8660254j, 0.0000000 + 0.0000000j],
```

Comment:
it is interesting that the first elements are different but second elements are the same between the rows. Seems to make sense but worth double checking the math.

Reply:
Yeah, I noticed that too. Here is the output from the pytorch fft on this tensor. I am only guessing it might be due to the nice ascending structure of the input tensor.

```
In [4]: tensor = torch.arange(12).reshape(4, 3)
   ...:    torch.fft.fft(tensor)
Out[4]:
tensor([[ 3.0000+0.0000j, -1.5000+0.8660j, -1.5000-0.8660j],
        [12.0000+0.0000j, -1.5000+0.8660j, -1.5000-0.8660j],
        [21.0000+0.0000j, -1.5000+0.8660j, -1.5000-0.8660j],
        [30.0000+0.0000j, -1.5000+0.8660j, -1.5000-0.8660j]])
        ```

Reply:
👍perhaps worth checking with numpy or scipy

Reply:
Better idea, taking a look!

Reply:
scipy also seems to agree with the results for this:
```
In [5]: from scipy.fft import fft
   ...: array = np.arange(12).reshape(4,3)
   ...: fft(array)
Out[5]:
array([[ 3. -0.j       , -1.5+0.8660254j, -1.5-0.8660254j],
       [12. -0.j       , -1.5+0.8660254j, -1.5-0.8660254j],
       [21. -0.j       , -1.5+0.8660254j, -1.5-0.8660254j],
       [30. -0.j       , -1.5+0.8660254j, -1.5-0.8660254j]])
```